lib/class-wp-theme-json-gutenberg.php

```diff
@@ -790,6 +791,78 @@ public function __construct( $theme_json = array( 'version' => WP_Theme_JSON_Gut
 		}
 	}
 
+	/**
+	 * Unwraps shared block style variations.
+	 *
+	 * It takes the shared variations (styles.variations.variationName) and
+	 * applies them to all the blocks that have the given variation registered
+	 * (styles.blocks.blockType.variations.variationName).
+	 *
+	 * For example, given the `core/paragraph` and `core/group` blocks have
+	 * registered the `section-a` style variation, and given the following input:
+	 *
+	 * {
+	 *   "styles": {
+	 *     "variations": {
+	 *       "section-a": { "color": { "background": "backgroundColor" } }
+	 *     }
+	 *   }
+	 * }
+	 *
+	 * It returns the following output:
+	 *
+	 * {
+	 *   "styles": {
+	 *     "blocks": {
+	 *       "core/paragraph": {
+	 *         "variations": {
+	 *             "section-a": { "color": { "background": "backgroundColor" } }
+	 *         },
+	 *       },
+	 *       "core/group": {
+	 *         "variations": {
+	 *           "section-a": { "color": { "background": "backgroundColor" } }
+	 *         }
+	 *       }
+	 *     }
+	 *   }
+	 * }
+	 *
+	 * @since 6.6.0
+	 *
+	 * @param array $theme_json A structure that follows the theme.json schema.
+	 * @return array Theme json data with shared variation definitions unwrapped under appropriate block types.
+	 */
+	private static function unwrap_shared_block_style_variations( $theme_json ) {
+		if ( ! isset( $theme_json['styles']['variations'] ) ) {
+			return $theme_json;
+		}
+
+		$registered_styles = WP_Block_Styles_Registry::get_instance()->get_all_registered();
```

Comment:
It doesn't appear that core block type variations (e.g. Button outline) get registered in the block styles registry.

We might be able to loop registered block types and check their `styles` property similar to what happens when determining valid variations for sanitization etc.

Reply:
Perhaps a better option would be to move the call to `unwrap_shared_block_style_variations` to after the collection of `$valid_variations`, then pass those as a new param. They should then contain both the core block styles as well as any registered in the `WP_Block_Styles_Registry`.

What do you think?

Reply:
In the interests of saving time, I've pushed a commit that I think does the job: https://github.com/WordPress/gutenberg/pull/62712/commits/bc758fce265e813bdd5b2c47a3f1cd8ae3147544.